lib/ClangImporter/ImportDecl.cpp

```diff
@@ -3415,6 +3415,36 @@ namespace {
       result = Impl.createDeclWithClangNode<StructDecl>(
           decl, AccessLevel::Public, Impl.importSourceLoc(decl->getBeginLoc()),
           name, Impl.importSourceLoc(decl->getLocation()), None, nullptr, dc);
+
+      // Check whether this record is trivial. During this check, we might
+      // discover that this record is unimportable, so it's important that this
+      // check happens before we cache the decl or import any of its members.
+      const clang::CXXRecordDecl *cxxRecordDecl =
+          dyn_cast<clang::CXXRecordDecl>(decl);
+      if (cxxRecordDecl) {
+        result->setIsCxxNonTrivial(!cxxRecordDecl->isTriviallyCopyable());
+
+        for (auto ctor : cxxRecordDecl->ctors()) {
+          if (ctor->isCopyConstructor()) {
+            // If we have no way of copying the type we can't import the class
+            // at all because we cannot express the correct semantics as a swift
+            // struct.
+            if (ctor->isDeleted() || ctor->getAccess() != clang::AS_public)
+              return nullptr;
+          }
+          if (ctor->getAccess() != clang::AS_public) {
+            result->setIsCxxNonTrivial(true);
+            break;
+          }
+        }
+
+        if (auto dtor = cxxRecordDecl->getDestructor()) {
+          if (dtor->isDeleted() || dtor->getAccess() != clang::AS_public) {
+            return nullptr;
+          }
+        }
+      }
```

Comment:
(I realize this PR is moving things around in some sense but:) From what I understand, this code is basically implementing an `isTrivial` method... is there a similar method in Clang which we can directly use/could you implement this upstream instead and use that here?

Even if this needs to be implemented here, it might be useful to have a separate function `isTrivial` (strawman name) which is called here, instead of it being implemented inline.

Reply:
We already have/use `cxxRecordDecl->isTriviallyCopyable()`. I don't know why we need L3436 as well. I will look into it.

Anyway, I will keep the call to `result->setIsCxxNonTrivial(!cxxRecordDecl->isTriviallyCopyable())` and add another function to determine whether the `cxxRecordDecl` is invalid (and we need to bail).